lib/pengine/rules.c

```diff
@@ -1032,7 +1032,11 @@ pe__eval_attr_expr(xmlNodePtr expr, pe_rule_eval_data_t *rule_data)
     if (value != NULL && h_val != NULL) {
         if (type == NULL) {
             if (pcmk__str_any_of(op, "lt", "lte", "gt", "gte", NULL)) {
-                type = "number";
+                if (pcmk__char_in_any_str('.', h_val, value, NULL)) {
+                    type = "number";
+                } else {
+                    type = "integer";
+                }
```

Comment:
I wonder if we should default to type="string" for lt, lte, gt, and gte if the value doesn't begin with a digit or decimal point, or whether that would be a compatibility concern.

I would think not a compatibility concern, because a user shouldn't expect any particular behavior from strings being compared as numbers.

Reply:
Note to self: Update/add regression tests if we do this.

Reply:
> I wonder if we should default to type="string" for lt, lte, gt, and gte if the value doesn't begin with a digit or decimal point, or whether that would be a compatibility concern.

Or a minus sign. I'm thinking a detect-text-type function returning 0 = string, 1 = integer, 2 = float. Or it might be easier to just try parsing doubles, and if that fails, try parsing integers, and if that fails it's a string. To be honest, I'd rather just default to string unconditionally, but that would break compatibility.

> I would think not a compatibility concern, because a user shouldn't expect any particular behavior from strings being compared as numbers.

I agree. The current behavior would be unintuitive.

Reply:
> Or it might be easier to just try parsing doubles, and if that fails, try parsing integers, and if that fails it's a string.

I'm just doing "try whichever numeric parse type was selected (if one was selected), and if that fails, then default to string." That will also be documented in Pacemaker Explained. Fixed in next push.

Reply:
I considered the "detect-text-type" function, but the "try X then default to Y on failure" seemed like a more straightforward approach. Since they're fairly lightweight, it's easier to let the parsing functions do our error checking for us instead of bloating the code with checks.